Add additional instructions for getting started (#1)
childmindresearch/niftyone: README.md

# NiftyOne

![NiftyOne Mosaic](.github/static/niftyone_mosaic_view.png)

NiftyOne is a platform for bulk visualization of large-scale neuroimaging datasets. It is built with [FiftyOne](https://docs.voxel51.com/).

- [NiftyOne](#niftyone)
  * [Installation](#installation)
  * [Testing](#testing)
  * [Usage](#usage)
  * [Running NiftyOne on a remote cluster](#running-niftyone-on-a-remote-cluster)
  * [FAQ](#faq)

## Installation

> **1. Install [conda](https://conda.io/projects/conda/en/latest/user-guide/install/index.html)**
> For stability, NiftyOne should be installed in it's own environment. For example,

Reviewer: ```suggestion
For stability, NiftyOne should be installed in it's own environment. For example, using [conda](https://docs.conda.io/projects/miniconda/en/latest/).
```

```bash
conda create -y -n niftyone python=3.10
conda activate niftyone

pip install -U pip
pip install git+https://github.com/cmi-dair/niftyone.git
```

> ## Testing

Reviewer: To me this feels a bit too in the weeds for the README. I suggest we keep the usage instructions minimal in the README and then move the more detailed tutorial content to a `TUTORIAL.md` file where we can freely go into details.

> 1. Install [datalad](https://www.datalad.org) to download the data files

Reviewer: Including a one-liner to install datalad that the user can copy-paste would be useful here.

```bash
conda install -n niftyone -c conda-forge datalad
```

> 2. You can use whatever BIDS dataset (such as [ds000102](https://openneuro.org/datasets/ds000102/versions/00001)  from OpenNeuro) as an example to test 

Reviewer: Ditto for getting the dataset from openneuro.

```bash
datalad install https://github.com/OpenNeuroDatasets/ds000102.git
cd ds000102
datalad get -r -J 8 .
```

```bash
niftyone bids_dir output_dir participant --workers 8

> ```

Reviewer: This is duplicated, should come later.

## Usage

> ### 1. Test launching the app

Reviewer: More description might be helpful here. Which app are we testing? Why?

```bash

> niftyone app launch

Reviewer: ```suggestion
fiftyone app launch
```

```

> ### 2. Generate figures for each participant

Reviewer: More description here as well. What does this do? What do the arguments mean?

```bash

> niftyone bids_dir output_dir participant --workers 8

Reviewer: It's nice if the user can just copy-paste the command.

```suggestion
niftyone ds000102 ds000102/derivatives/niftyone participant --workers 8
```

```

> ### 3. Collect participant figures into a FiftyOne dataset

Reviewer: Ditto, more description, make command copy-pasteable. Here and below.

```bash
niftyone bids_dir output_dir group
```

> ### 4. Launch FiftyOne app

Reviewer: How do I then open the app? Would be useful to add here the expected console output:

```
App launched. Point your web browser to http://localhost:5151

To exit, close the App or press ctrl + c
```

And add a screenshot for the app loaded in the browser.

```bash
niftyone bids_dir output_dir launch
```

> ## Running NiftyOne on a remote cluster

Reviewer: This gets a little in the weeds. I might consolidate into the above steps as **5. Running NiftyOne remotely**. Then just explain how you can launch niftyone remotely as usual and forward the port.  

You can assume a simple remote server, i.e. not a cluster requiring compute node reservation, multi-hop forward, etc. Pointers for how to adapt to the cluster/multi-hop case can come at the end.

 
### 4) Port Forwarding

Open a second terminal.

Execute the following command to create a multihop port forward from your local machine --> login node --> compute node. Related [gist](https://gist.github.com/clane9/ea3a469e727b6e7f75dc4373e9d2241d)

  
HOST=(name of compute node, not the full address. e.g. r001)
PORT=5151
```
ssh -L $PORT:localhost:$PORT $USER@server.com ssh -L $PORT:localhost:$PORT -N $HOST
```
  

### 5) Open [http://localhost:5151](http://localhost:5151) on your local browser to check the app.

  

### Additional reference: 

https://superuser.com/questions/1154383/ssh-port-forwarding-on-windows

https://phoenixnap.com/kb/ssh-port-forwarding
## FAQ

### Q1) Where can NiftyOne be installed?

NiftyOne can be installed both locally and on a remote server. If NiftyOne is installed on a server, the output can be viewed from a local web browser after port fowarding.

  

### Q2) Port forwarding is not working and I am getting connection errors

Try testing with a simple HTTP server with a "Hello World" HTML file using port fowarding to rule out any networking problems before installing further software.

  

> $python3 -m http.server 9000

Reviewer: Formatting.

https://www.digitalocean.com/community/tutorials/python-simplehttpserver-http-server#python-simplehttpserver-error-no-module-named-simplehttpserver

  
 
###  Q3) I am getting the following error:

> /home/user/.../mongod: error while loading shared libraries

Reviewer: Formatting

  

Try re-installing mongo and re-creating the virtual conda environment.

  

###  Q4) I am getting the following error:

> OSError: You must have fiftyone>=0.22.1 installed in order to migrate from v0.22.1 to v0.21.6, but you are currently running fiftyone==0.21.6.

Reviewer: Formatting
